source/EventBus.Core.pas

```diff
@@ -304,30 +305,36 @@ procedure TEventBus.RegisterSubscriber<T>(ASubscriber: TObject; ARaiseExcIfEmpty
   try
     LSubscriberClass := ASubscriber.ClassType;
     LSubscriberMethods := TSubscribersFinder.FindSubscriberMethods<T>(LSubscriberClass, ARaiseExcIfEmpty);
-    for LSubscriberMethod in LSubscriberMethods do Subscribe<T>(ASubscriber, LSubscriberMethod);
+
+    for LSubscriberMethod in LSubscriberMethods do begin
+      // If the method do not have its Context specified, we assign the instance-specific context to it.
+      // Note - the instance-specific context itself may be empty.
+      if LSubscriberMethod.Context = '' then LSubscriberMethod.SetNewContext(AInstanceContext);

```

Comment:
here we have to think who is the master... If I have set something using the Attribute I guess I want to override it with the InstanceContext. So it becomes:
```
if not AInstanceContext = '' then LSubscriberMethod.SetNewContext(AInstanceContext);
```
What do you think? My idea is that the runtime context should win over the attribute one

Reply:
@spinettaro @dkounal
Yes.  If the subscriber method doesn't have its Context specified through compile time attribute (i.e., an empty string), then we will use instance context (which itself could be an empty string too) as the method's context.  But the point is:
1. Runtime context should win over
2. We use - "whether the compile time attribute has explicitly specified the Context as a non-empty string" - as an indicator for accepting runtime instance context or not.

Reply:
The most flexible is to be able to override the context after compile time.

Reply:
> The most flexible is to be able to override the context after compile time even if it is not empty

Yes. Agree. 
So this line ` if LSubscriberMethod.Context = '' then LSubscriberMethod.SetNewContext(AInstanceContext);`  actually achieves the following logic:
1. The user specifies compile-time attributes.  Some attributes leaves Context empty
2. At runtime,  those empty compile time attributes will be overwritten by runtime instance context.  Note - the runtime instance context itself can be empty. But as long as the compile time context is empty, runtime instance context will be set as the method's context
3. This way,  the run-time context will only be applied to those methods whose compile time context is empty, thus giving us the most flexibility, rather than using run-time context to replacing ALL compile time context.

Reply:
> 
> 
> The most flexible is to be able to override the context after compile time even if it is not empty.
> The less flexible way is to define empty context in attributes as context that could be set at runtime (instance based).

I see your points.  I am good with it,  hence the line  `if not AInstanceContext = '' then LSubscriberMethod.SetNewContext(AInstanceContext);
`
I am OK either way, as long as this behavior is clearly documented.

Reply:
I have no problem with any of these two cases.

Reply:
@spinettaro has right, a exception should be fired if context is not empty by attribute and an instance based context is provided

Reply:
@spinettaro 
`if LSubscriberMethod.Context = '' then LSubscriberMethod.SetNewContext(AInstanceContext);`  --- would result in "**_selective context replacemen_**t", meaning runtime context (whatever it is) will be selectively applied to those methods whose compile time context is unspecified (i.e., empty str)

`if not AInstanceContext = '' then LSubscriberMethod.SetNewContext(AInstanceContext);`  --- would result in "**_batch context replacement_**", meaning runtime context (if it is not empty) will be applied for ALL subscriber methods regardless of the compile-time context.

I think "selective replacement" allows the user to specify at compile time which methods is eligible for runtime context. Hence more flexible. 

What do you think?

Reply:
@wxinix oh I see. Sorry! The if statement is inside the loop, I need to think about it because as you pointed you need a "selective" action. We should be able from outside to specifically declare context for a specific event. Not for all of them

Reply:
@spinettaro  That was the reason I did it with defining a context value that the Subscriber method will replace it with the new context, if a new context was provided during subscribing

Reply:
> That was the reason I did it with defining a context value
Yes.  Either use "EmptyStr" or a "special" context value to indicate the compile-time context is eligible for runtime replacement.  To that end,  probably"EmptyStr" is easier and straightforward?

Reply:
> 
> 
> > That was the reason I did it with defining a context value
> 
> Yes. Either use "EmptyStr" (which is what PR #58 does) or a "special" context value to indicate that the compile-time context is eligible for runtime replacement. To that end, probably"EmptyStr" is easier and straightforward?

The best is what you proposed yesterday, to have an extra attribute argument to state that this context is instance based

Reply:
I think EmptyStr is better, otherwise with the other one we have also to handle the logic in case an  "InstanceBased" is declared as context but never overridden. In general I have the feeling we don't have the control of it, overriding everything because is empty string makes me a little bit nervous. A better solution would be the ability to specify the event and the context. 
what about a new function **RegisterNewContext**? where you have to pass the Subscriber (TObject), the EventType (the interface) and context (the new context string). we can find for the pair Subscriber-EventType into our registrations and then change the context of that specific event, in that particular Subscriber.
Thoughts? @wxinix @dkounal

Reply:
> what about a new function **RegisterNewContext**

I had thought of having a similar function RegisteInstanceContext, as a method of a new interface ISubscriberRegistration, which is returned by IEventBus.RegisterSubscriberForEvent: ISubscriberRegistration, then a user can do a fluent call:

RegisterSubscriberForEvent.RegisterInstanceContext('NewInstanceContext');

This seems a more elegant solution.

Reply:
I believe we make things very complex.
A subscriber can have one or more subscriptions and we need to mark which subscription will allow a context override or not. This is known before compiling. 
If all setup is based till now in attributes, why not extending the attributes we use to mark this subscription?

Reply:
> I believe we make things very complex

Yes - it might be an overkill, making things overly complicated.